docs/source/examples/word-count.rst

```diff
@@ -0,0 +1,247 @@
+   /us/> . To provide attribution, please cite to "ZL Technologies, Inc. (http:
+   //www.zlti.com)."\r\n***********\r\nDate: Wed, 29 Nov 2000 09:40:00 -0800 (P
+   ST)\r\nFrom: Jill T Zivley\r\nTo: Robert Cook, Robert Crockett, John Handley
+   , Shawna'
+
+Create a function to count words in each file:
+
+.. code-block:: python
+
+   >>> def count_words(fn):
+   ...     word_counts = defaultdict(int)
+   ...     with hdfs.open(fn) as f:
+   ...         for line in f:
+   ...             for word in line.split():
+   ...                 word_counts[word] += 1
+   ...     return word_counts
+
+Count the number of words in the first text file locally:
+
+.. code-block:: python
+
+   >>> counts = count_words(filenames[0])
+   >>> print(sorted(counts.items(), key=lambda k_v: k_v[1], reverse=True)[:10])
+
+   [(b'the', 144873),
+    (b'of', 98122),
+    (b'to', 97202),
+    (b'and', 90575),
+    (b'or', 60305),
+    (b'in', 53869),
+    (b'a', 43300),
+    (b'any', 31632),
+    (b'by', 31515),
+    (b'is', 30055)]
+
+Count the number of words in the first text file on a ``distributed`` worker:
+
+.. code-block:: python
+
+   >>> future = e.submit(count_words, filenames[0])
+   >>> counts = future.result()
+   >>> print(sorted(counts.items(), key=lambda k_v: k_v[1], reverse=True)[:10])
+
+   [(b'the', 144873),
+    (b'of', 98122),
+    (b'to', 97202),
+    (b'and', 90575),
+    (b'or', 60305),
+    (b'in', 53869),
+    (b'a', 43300),
+    (b'any', 31632),
+    (b'by', 31515),
+    (b'is', 30055)]
+
+Count the number of words in all of the text files using ``distributed``
+workers. Note that the ``map`` operation is non-blocking, and you can continue
+to work in the Python shell/notebook.
+
+.. code-block:: python
+
+   >>> futures = e.map(count_words, filenames)
+
+Check the status of some ``futures`` while all of the text files are being
+processed:
+
+.. code-block:: python
+
+   >>> len(futures)
+
+   161
+
+   >>> futures[:5]
+
+   [<Future: status: finished, key: count_words-5114ab5911de1b071295999c9049e941>,
+    <Future: status: pending, key: count_words-d9e0d9daf6a1eab4ca1f26033d2714e7>,
+    <Future: status: pending, key: count_words-d2f365a2360a075519713e9380af45c5>,
+    <Future: status: pending, key: count_words-bae65a245042325b4c77fc8dde1acf1e>,
+    <Future: status: pending, key: count_words-03e82a9b707c7e36eab95f4feec1b173>]
+
+   >>> progress(futures)
+
+   [########################################] | 100% Completed |  3min  0.2s
+
+When the ``futures`` finish reading in all of the text files and counting
+words, the results will exist on each worker. This operation required about
+3 minutes to run on a cluster with three worker machines, each with 4 cores
+and 16 GB RAM.
+
+To sum the word counts for all of the text files, we can iterate over the
+``futures`` and update a local dictionary that contains all of the word counts.
+This operation requires a few minutes as it pulls the data from each worker
+to the local process.
```

Comment:
It's worth saying that this is not the most efficient way to accomplish this work.

Another option to speed all of this up is to cull the dictionaries before gathering them.  Perhaps we only need the top 1000 words from each file.  This would be easy to compute on the worker, would demonstrate mapping onto futures, and would significantly decrease communication costs.